ext/web/08_text_encoding.js

```diff
@@ -91,40 +94,43 @@
           allowShared: true,
         });
       }
-      options = webidl.converters.TextDecodeOptions(options, {
-        prefix,
-        context: "Argument 2",
-      });
+      let stream = false;
+      if (options !== undefined) {
+        options = webidl.converters.TextDecodeOptions(options, {
+          prefix,
+          context: "Argument 2",
+        });
+        stream = options.stream;
+      }
 
       try {
-        try {
-          if (ArrayBufferIsView(input)) {
-            input = new Uint8Array(
-              input.buffer,
-              input.byteOffset,
-              input.byteLength,
-            );
-          } else {
-            input = new Uint8Array(input);
-          }
-        } catch {
-          // If the buffer is detached, just create a new empty Uint8Array.
+        // Note from spec: implementations are strongly encouraged to use an implementation strategy that avoids this copy.
+        // When doing so they will have to make sure that changes to input do not affect future calls to decode().
+        if (input.byteLength === 0) {
+          // This handles the case where buffer is detached too.
           input = new Uint8Array();
         }
-        if (
-          ObjectPrototypeIsPrototypeOf(
-            SharedArrayBuffer.prototype,
-            input.buffer,
-          )
-        ) {
-          // We clone the data into a non-shared ArrayBuffer so we can pass it
-          // to Rust.
-          // `input` is now a Uint8Array, and calling the TypedArray constructor
-          // with a TypedArray argument copies the data.
-          input = new Uint8Array(input);
-        }
 
-        if (!options.stream && this.#rid === null) {
+        // if (
+        //   ObjectPrototypeIsPrototypeOf(
+        //     SharedArrayBuffer.prototype,
+        //     input.buffer,
+        //   )
+        // ) {
+        //   // We clone the data into a non-shared ArrayBuffer so we can pass it
+        //   // to Rust.
+        //   // `input` is now a Uint8Array, and calling the TypedArray constructor
+        //   // with a TypedArray argument copies the data.
+        //   input = new Uint8Array(input);
+        // }
```

Comment:
I'm not sure why we did this but passing SABs to Rust works without copying into a non-SAB

Reply:
It used to be the case that passing SABs to Rust didn't work, because `ZeroCopyBuf` explicitly disallowed SABs. #15922 made this possible, and apparently no one noticed.

Accessing SAB data as a `&[u8]` or `&mut [u8]` is unsound because the buffer could be modified at any point by a worker. It has essentially the same safety guarantees as transmuting a `&[AtomicU8]`.

Reply:
ok, i'll can explict copying of SABs in the op macro codegen for `&[u8]`. that works?

Reply:
Sure.